make: Support publishing custom images to private registries

barkardk opened the pull request:
To make it easier to build custom images and deploy from private registries  its nice to be able to push via makefile without having to use buildx. 
Remove the hardcoded latest tagging from the docker image to a configurable tag with value:latest set to default instead.

olix0r commented:
How does differ from the `VERSION` variable and `make push`?

https://github.com/linkerd/linkerd2-proxy-init/blob/51a35572729e3eb2bfff96ec20aaaf3a857deac3/Makefile#L60-L70

barkardk commented:
I wanted to make use of docker build and docker push without using docker buildx. 
I assumed docker buildx was there for pipeline usages and did not want to ask for its removal.  
 I get the error 
```multiple platforms feature is currently not supported for docker driver. ``` 
when I try to use the `make images`  or `make push` command. 
I just thought is easier to stick to the   
```make image```   as is noted in the README with a custom tag and then just add an entry for a simple docker push

olix0r commented:
@barkardk makes sense. Do we need both `TAG` and `VERSION`?

barkardk commented:
@olix0r not sure originally `make image` had latest as hardcoded but VERSION uses git revision. 
```diff
image: ## Build docker image for the project
	DOCKER_BUILDKIT=1 docker build -t $(REPO):latest .
```
I set a new param TAG instead of using VERSION to preserve backwards compatibility.  
Using VERSION makes perfect sense to me if no one has an issue with replacing the default `latest` tag

barkardk commented:
@olix0r I took out the TAG parameter and replaced it with version.  
 Also added some documentation and combined the make commands so  that docker-push will also build the container.  To avoid the extra step

olix0r commented on Makefile:
```diff
@@ -42,7 +42,13 @@ integration-test: image ## Perform integration test
-	DOCKER_BUILDKIT=1 docker build -t $(REPO):latest .
+	DOCKER_BUILDKIT=1 docker build -t $(REPO):$(TAG) .
+
+.PHONY: docker-push
+docker-push: ## push to any docker registry
+	make image
+	docker push $(REPO):$(TAG)
```
I'm a little concerned about recursively invoking `make`... I know this can be fragile.

Can we accomplish the same thing with:
```suggestion
# push to any docker registry
.PHONY: docker-push
docker-push: image
	docker push $(REPO):$(TAG)
```

olix0r commented on Makefile:
```diff
@@ -42,7 +42,13 @@ integration-test: image ## Perform integration test
 .PHONY: image
 image: ## Build docker image for the project
-	DOCKER_BUILDKIT=1 docker build -t $(REPO):latest .
+	DOCKER_BUILDKIT=1 docker build -t $(REPO):$(TAG) .
```
perhaps the makefile should declare `DOCKER_BUILDKIT = 1` globally above?

olix0r reviewed (commented):
Thanks for humoring me ;)

kleimkuhler commented:
Hey @barkardk just checking in on this. Do you still have the interest/time to address the remaining comments on this and get it merged?

alpeb commented:
Closing this one for lack of activity. @barkardk you're welcomed to reopen if you're still interested.
